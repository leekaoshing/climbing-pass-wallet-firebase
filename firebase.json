{
  "storage": {
    "rules": "storage.rules"
  },
  "emulators": {
    "firestore": {
      "port": 9091,
      "host": "0.0.0.0"
    },
    "hosting": {
      "port": 5000,
      "host": "0.0.0.0"
    },
    "functions": {
      "port": 5001
    },
    "auth": {
      "port": 9099,
      "host": "0.0.0.0"
    },
    "ui": {
      "enabled": true,
      "port": 4000,
      "host": "0.0.0.0"
    }
  },
  "firestore": {
    "rules": "firestore.rules",
    "indexes": "firestore.indexes.json"
  },
  "hosting": {
    "site": "climbingpasswallet",
    "public": "build",
    "predeploy": "npm run build",
    "ignore": [
      "firebase.json",
      "**/.*",
      "**/node_modules/**"
    ],
    "rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      }
    ]
  },
  "functions": {
    "predeploy": [
      "npm --prefix \"$RESOURCE_DIR\" run lint"
    ],
    "ignore": [
      "**/.*",
      "**/*.spec.js",
      "**/node_modules/**",
      "jest.config.js",
      "coverage/**",
      "src",
      "scripts/**",
      "jsconfig.json"
    ],
    "source": "functions"
  },
  "remoteconfig": {
    "template": "remoteconfig.template.json"
  }
}
