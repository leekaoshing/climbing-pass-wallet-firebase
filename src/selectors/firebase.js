import { isLoaded, isEmpty } from 'react-redux-firebase';

export const selectLoggedInUser = state => {
    const auth = state.firebase.auth;
    return isLoaded(auth) && !isEmpty(auth) && state.firestore.data.users && state.firestore.data.users[auth.uid];
}

export const selectUsers = state => {
    return state.firestore.data && state.firestore.data.users;
}

export const selectFirebaseAuth = state => {
    return state.firebase.auth;
}

export const selectGyms = state => {
    return state.firestore.data.gyms;
}