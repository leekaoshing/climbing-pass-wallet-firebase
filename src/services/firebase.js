// Firebase App (the core Firebase SDK) is always required and
// must be listed before other Firebase SDKs
import firebase from 'firebase/app';

// Add the Firebase services that you want to use
// We only want to use Firebase Auth here
import 'firebase/auth';
import 'firebase/firestore';

// Your app's Firebase configuration
const firebaseConfig = {
    apiKey: 'AIzaSyD3WTY0d0oHdZSxnAghqg-F7hKJ74jrtyA',
    authDomain: 'climbing-pass-wallet.firebaseapp.com',
    projectId: 'climbing-pass-wallet',
    storageBucket: 'climbing-pass-wallet.appspot.com',
    messagingSenderId: '287194549775',
    appId: '1:287194549775:web:b4792b4fd172931caf9d7a',
    measurementId: 'G-ZLTF129GET'
};

// Initialize Firebase
firebase.initializeApp(firebaseConfig);

export const auth = firebase.auth();
export const firestore = firebase.firestore();
if (window.location.hostname === "localhost") {  // Emulator
    firestore.useEmulator("localhost", 9091);
    auth.useEmulator('http://localhost:9099/');
}

// Google sign in
const provider = new firebase.auth.GoogleAuthProvider();
export const signInWithGoogle = () => {
    auth.signInWithPopup(provider);
};

export const generateUserDocument = async (user, additionalData) => {
    if (!user) return;
    console.log('generate user document', additionalData);
    console.log('generate user document user', user);
    const userRef = firestore.doc(`users/${user.uid}`);
    const snapshot = await userRef.get();

    if (!snapshot.exists) {
        // try {
            await userRef.set({
                ...additionalData,
                passes: {}
            });
        // }
        // catch (error) {
        //     throw error;  // TODO handle here, throw to user
        // }
    }
    return getUserDocument(user.uid);
};

// export const getEmailFromDisplayName = async (displayName) => {
//     if (!displayName) return;

//     const usersCollectionRef = firestore.collection('users');
//     const user = await usersCollectionRef.where('displayName', '==', displayName) // TODO Problem: not allowed to look for users if user is not signed in, since users will only be allowed to see their own data
//         .get()
//         .then(results => {
//             if (!results.empty) {
//                 console.error('user does not exist!') // TODO handle here, throw to user
//             }
//             return null;
//         })
//         .catch(error => {
//             console.error('Unable to check if user \'' + additionalData.displayName + '\' exists: ', error); // TODO handle here, throw to user
//     })

//     // if (user !== null) {
//     //     auth.getUserFromUid(user.uid) //.... continue this
//     //     return Email;
//     // }
// }

export const getUser = async (user) => {
    if (!user) return;
    return getUserDocument(user.uid);
}

export const isDisplayNameTaken = async (displayName) => {
    if (!displayName) return;
    return await firestore.collection('users').where('displayName', '==', displayName)
        .get()
        .then(results => {
            if (!results.empty) {
                return true;
            }
            return false;
        })
}

const getUserDocument = async uid => {
    if (!uid) return null;
    try {
        const userDocument = await firestore.doc(`users/${uid}`).get();
        return {
            uid,
            ...userDocument.data()
        };
    } catch (error) {
        throw error;  // TODO handle here, throw to user
    }
};

export const signOut = () => {
    firebase.auth().signOut().then(() => {
        // Sign-out successful.
      }).catch((error) => {
        // An error happened.
      });
}

// Finally, export it to use it throughout your app
export default firebase;