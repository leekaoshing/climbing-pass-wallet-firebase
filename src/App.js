import { CircularProgress } from '@material-ui/core';
import Container from '@material-ui/core/Container';
import { default as React, useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { isEmpty, isLoaded, useFirestoreConnect } from 'react-redux-firebase';
import { useCheckMobileScreen } from './actions/actions';
import './App.css';
import { AboutDialog } from './components/AboutDialog';
import { Login } from './components/Login';
import { NavBar } from './components/NavBar';
import { Passes } from './components/Passes';
import { SignUpDialog } from './components/SignUpDialog';
import { selectGyms, selectLoggedInUser, selectUsers } from './selectors/firebase';

const getScreenSize = isMobile => {
  if (isMobile) {
    return 'xs';
  } else {
    return 'sm';
  }
}

function App() {
  const auth = useSelector((state) => state.firebase.auth);

  const isScreenSizeMobile = useCheckMobileScreen();

  const [loading, setLoading] = useState(true);
  const [showUser, setShowUser] = useState(false);
  const users = useSelector(selectUsers);

  useFirestoreConnect([
    {
      collection: 'users'
    },
    {
      collection: 'gyms'
    }
  ])
  const gyms = useSelector(selectGyms);

  const loggedInUser = useSelector(selectLoggedInUser);
  useEffect(() => {
    setLoading(!isLoaded(auth) || (!isEmpty(auth) && (!users || !Object.keys(users).includes(auth.uid))));
    setShowUser(loggedInUser);
  }, [auth, users, loggedInUser])

  return (
    <div className="App" style={{ backgroundColor: 'white', height: '100vh', minHeight: '100vh' }}>
      <NavBar />
      <Container maxWidth={getScreenSize(isScreenSizeMobile)}>

        <AboutDialog />
        <SignUpDialog />
        {/* {
          users && users[auth.uid] ?
            <Passes />
            :
            <Login />
        } */}

        {
          loading ?
            <div style={{ display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
              <br /><br /><br /><br /><br />
              <CircularProgress />
            </div>
            :
            showUser ?
              <Passes />
              :
              <Login />
        }
        <br />
        <br />
      </Container>
    </div>
  );
}

export default App;