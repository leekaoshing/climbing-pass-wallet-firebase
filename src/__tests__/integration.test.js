import { screen, waitFor } from '@testing-library/dom';
import { fireEvent, render, act, waitForElement } from '@testing-library/react';
import firebase from 'firebase/app';
import React from 'react';
import { Provider } from 'react-redux';
import { ReactReduxFirebaseProvider } from 'react-redux-firebase';
import { createFirestoreInstance } from 'redux-firestore';
import App from '../App';
import importedStore, { rrfConfig } from '../app/store';
import '../index.css';

const rrfProps = {
    firebase,
    config: rrfConfig,
    dispatch: importedStore.dispatch,
    createFirestoreInstance
};

const FullApp = ({ store }) => {
    return (
        <Provider store={store}>
            <ReactReduxFirebaseProvider {...rrfProps}>
                <App />
            </ReactReduxFirebaseProvider>
        </Provider>
    );
}

jest.setTimeout(10000);
jest.disableAutomock()

describe.skip('Full end-to-end integration test', () => {
    jest.unmock('react-redux-firebase');

    // let appRender;

    // beforeAll(() => {
    //     appRender = render(
    //         <FullApp store={importedStore} />
    //     )
    //     // await act(async () => {
    //     //     appRender = mount(<FullApp store={importedStore} />);
    //     // });
    // })

    // it('Should render correctly', async () => {
    //     await screen.findAllByText(/Email/i)
    //     expect(appRender.asFragment()).toMatchSnapshot();
    // })

    it('Should login correctly', async () => {
        // appRender.rerender(<FullApp store={importedStore} />);
        // expect(appRender.asFragment()).toMatchSnapshot();

        // const appRender = render(
        //     <FullApp store={importedStore} />
        // )

        const { rerender } = render(
            <Provider store={importedStore}>
                <ReactReduxFirebaseProvider {...rrfProps}>
                    <App />
                </ReactReduxFirebaseProvider>
            </Provider>
            );


        await screen.findAllByText(/Email/i)

        fireEvent.change(screen.getByLabelText('Email'), { target: { value: 'rban@mail.com' } });
        fireEvent.change(screen.getByLabelText('Password'), { target: { value: '123456' } });
        fireEvent.click(screen.getByText(/Log In/i));
        // rerender(<FullApp store={importedStore} />, container);

        
        await waitFor(() => expect(importedStore.getState().firebase.auth.isEmpty).toBe(false));
        
        
        rerender(<Provider store={importedStore}>
            <ReactReduxFirebaseProvider {...rrfProps}>
                <App />
            </ReactReduxFirebaseProvider>
        </Provider>)
        console.log('STATE', importedStore.getState())
        await waitFor(() => expect(importedStore.getState().firestore.data['gyms']).toBeDefined());
        const username = await screen.findByLabelText(/Ray Ban/i);

        // await waitForElementToBeRemoved(() => screen.queryByLabelText('Email')).then(() => {
        //     console.log(importedStore.getState());
        //     // console.log('DONE HERE');
        //     screen.findByText(/Ray Ban/i).then(username => {
        //         expect(username).toBeInTheDocument();
        //     }).catch(err => {
        //         console.log('FAILED', err)
        //     })

        //     expect(appRender.asFragment()).toMatchSnapshot();
        // }).catch(err => {
        //     console.log('FAILED', err)
        // })
        // await screen.findAllByText(/Ray Ban/i)
        // const username = await screen.findByText(/Ray Ban/i)
        // expect(username).toBeInTheDocument();
        // appRender.update();
        // expect(appRender.asFragment()).toMatchSnapshot();
    })


});