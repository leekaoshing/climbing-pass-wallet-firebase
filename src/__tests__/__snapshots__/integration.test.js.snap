// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Full end-to-end integration test Should login correctly 1`] = `
<DocumentFragment>
  <div
    class="App"
    style="background-color: white; height: 100vh; min-height: 100vh;"
  >
    <div
      class="makeStyles-root-1"
    >
      <header
        class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation0"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
          style="display: flex; align-items: center; justify-content: center;"
        >
          <button
            aria-label="menu"
            class="MuiButtonBase-root MuiIconButton-root makeStyles-homeButton-2 MuiIconButton-colorInherit MuiIconButton-edgeStart"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <p
            class="MuiTypography-root makeStyles-title-3 MuiTypography-body2"
          >
            Climbing Pass Wallet
          </p>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-colorInherit MuiButton-outlinedSizeSmall MuiButton-sizeSmall"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              About
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </header>
    </div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <div />
      <div
        style="display: flex; align-items: center; justify-content: center;"
      >
        <br />
        <br />
        <br />
        <br />
        <br />
        <div
          class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 40px; height: 40px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </div>
      <br />
      <br />
    </div>
  </div>
</DocumentFragment>
`;
